### PR TITLE
Use ENU convention

### PR DESCRIPTION
**Pull request by mistoll**
Here's an ENU implementation of the the algorithm (#36).
There's also a matrix implementation of the initial orientation computation, which should be more stable than the RPY implementation.

I removed the micro optimizations, hoping the compiler does that job. The "static inline" methods should force compiler optimization while keeping the code easy to read and maintain.

Not tested yet, only for review.

What's missing:
- Switch between ENU / NED by parameter
- Provide computeOrientation in reusable library

### CODE REVIEW

**Comment by mintar** [comment]
## Thanks, I'll look into it next week!

Diese Nachricht wurde von meinem Android-Mobiltelefon mit K-9 Mail gesendet.

**Comment by mistoll** [comment]
Great. 
I'm still fixing some bugs. Just give me some feedback on the style and the optimizations, please.

**Comment by mintar** [comment]
Ok, I'm looking at it right now and will comment as I go. Sorry for being brief. :)

**Comment by mintar** [comment]
This branch should be rebased (`git rebase -i`) onto `indigo`, while dropping 
035eaa7adcb16e74cc54875fd5ddd9f14bfca6b3 and 3f878dc91026b3a21bcc5ecb5406f4e1b88df072. It took me a while to figure out that the combination of those two commits is a no-op. Afterwards you can just force-push to your feature branch.

**Comment by mintar** [comment]
About the "removed micro optimizations": I guess you are talking about the `union` in `invSquareRoot()`, right? That actually is there for a reason: #17 ; you've just reintroduced that bug. So I would prefer you left the `invSqrt()` function untouched, including the comments.

That said, I'm open to removing the whole `invSqrt()` function altogether and replacing it with `1 / sqrt()`. But let's do that in a separate pull request; this one is hard enough to review as it is because it's so big.

**Comment by mintar** [comment]
Wow, that was quick. :)

File: imu_filter_madgwick/CMakeLists.txt [hunk 1]
```diff
@@ -24,12 +24,13 @@ include_directories(
 
 
 # create imu_filter library
-add_library (imu_filter src/imu_filter.cpp  src/imu_filter_ros.cpp  include/imu_filter_madgwick/imu_filter.h)
+set(CMAKE_CXX_FLAGS "-O2")
```
Comment (mintar) [inline imu_filter_madgwick/CMakeLists.txt]: You shouldn't override `CMAKE_CXX_FLAGS`. First, the CMakeLists.txt is the wrong place to do this; if you want a release build, do something like `catkin_make -DCMAKE_BUILD_TYPE=Release` instead. Second, the original contents of `CMAKE_CXX_FLAGS` are lost, which can break the whole catkin workspace or break the build on different architectures. Just remove that line.

**Comment by mistoll** [comment]
The no-op commits should be removed. The commits of this pull request are somewhat mixed up and the later revert changes of the predecessors. If you like, I will squash them all together.

With "micro optimizations" I mean the explicit computation of terms, which are used multiple times (" Auxiliary variables to avoid repeated arithmetic").

I copied the invSqrt from the wrong place. Didn't want to change that. I tried using 1/sqrt, but it raises computation time about 10-20% (gtest run-time with 100000 iterations)

File: imu_filter_madgwick/CMakeLists.txt [hunk 2]
```diff
@@ -52,4 +53,17 @@ install(DIRECTORY include/${PROJECT_NAME}/
 
 install(FILES imu_filter_nodelet.xml
   DESTINATION ${CATKIN_PACKAGE_SHARE_DESTINATION}
+)
+
+
+
+if(CATKIN_ENABLE_TESTING)
+  catkin_add_gtest(${PROJECT_NAME}-madgwick_test
```
Comment (mintar) [inline imu_filter_madgwick/CMakeLists.txt:60]: That macro is defined in the package `rosunit`. You should add a line `<test_depend>rosunit</test_depend>` to the `package.xml`

**Comment by mintar** [comment]
> The no-op commits should be removed. The commits of this pull request are somewhat mixed up and the later revert changes of the predecessors. If you like, I will squash them all together.

No, that's fine. Let's do that in the end, when we're happy with the result. Better just keep adding commits for now, that makes it easier for me to track which commits I've already reviewed.

> With "micro optimizations" I mean the explicit computation of terms, which are used multiple times (" Auxiliary variables to avoid repeated arithmetic").

Ok, that's a good idea; you're right, I expect any modern compiler to do that. I think they came originally from Sebastian Madgwick's implementation, which is intended for very small embedded chips that don't even have floating point arithmetic (so the original code used integer arithmetic everywhere).

> I copied the invSqrt from the wrong place. Didn't want to change that. I tried using 1/sqrt, but it raises computation time about 10-20% (gtest run-time with 100000 iterations)

Ah, good to know. So let's keep `invSqrt`.

File: imu_filter_madgwick/include/imu_filter_madgwick/imu_filter.h [hunk 3]
```diff
@@ -71,6 +53,11 @@ class ImuFilter
         zeta_ = zeta;
     }
 
+    void setUseNED(bool useNED)
+    {
+	useNED_ = useNED;
```
Comment (mintar) [inline imu_filter_madgwick/include/imu_filter_madgwick/imu_filter.h]: This line is indented with a tab; use spaces instead. (and sorry for nitpicking)

File: imu_filter_madgwick/src/imu_filter.cpp [hunk 4]
```diff
@@ -25,10 +25,146 @@
...
+{
+  // Rate of change of quaternion from gyroscope
+  // See EQ 12
+  qDot1 = 0.5f * (-q1 * gx - q2 * gy - q3 * gz);
+  qDot2 = 0.5f * (q0 * gx + q2 * gz - q3 * gy);
+  qDot3 = 0.5f * (q0 * gy - q1 * gz + q3 * gx);
+  qDot4 = 0.5f * (q0 * gz + q1 * gy - q2 * gx);
+}
+
+static inline void addGradientDecentStep(
```
Comment (mintar) [inline imu_filter_madgwick/src/imu_filter.cpp]: Typo: decent => descent

File: imu_filter_madgwick/src/imu_filter.cpp [hunk 5]
```diff
@@ -37,195 +173,125 @@ ImuFilter::~ImuFilter()
...
+    normalizeVector(mx, my, mz);
+
+    // Compensate for magnetic distortion
+    compensateMagneticDistortion(q0, q1, q2, q3, mx, my, mz, _2bxy, _2bz);
+
+    // Gradient decent algorithm corrective step
+    s0 = 0.0;  s1 = 0.0;  s2 = 0.0;  s3 = 0.0;
+    if (useNED_) {
+      // Not optimized version
+      // This code runs as fast as the optimized version when compiled with -O2
```
Comment (mintar) [inline imu_filter_madgwick/src/imu_filter.cpp]: I'd remove those two comment lines; this makes sense in the commit message, but for somebody reading this code for the first time, it's not clear what the "optimized version" is.

File: imu_filter_madgwick/test/madgwick_test.cpp [hunk 6]
```diff
@@ -0,0 +1,115 @@
...
+  }
+
+  filter.getOrientation(q0,q1,q2,q3);
+  ASSERT_LT(fabs(q1), MAX_DIFF);
+  ASSERT_LT(fabs(q2), MAX_DIFF);
+  ASSERT_LT(fabs(q3), MAX_DIFF);
+}
+
+
+TEST(StaticOrientationTest, Stationary2){
```
Comment (mintar) [inline imu_filter_madgwick/test/madgwick_test.cpp]: StaticOrientationTest => MadgwickTest

File: imu_filter_madgwick/test/madgwick_test.cpp [hunk 7]
```diff
@@ -0,0 +1,115 @@
...
+      filter.madgwickAHRSupdate(Gx, Gy, Gz, Ax, Ay, Az, Mx, My, Mz, dt);
+  }
+  filter.getOrientation(q0,q1,q2,q3);
+
+  ASSERT_LT(fabs(q1 * q0 - 0.25), MAX_DIFF);
+  ASSERT_LT(fabs(q2 * q0 + 0.25), MAX_DIFF);
+  ASSERT_LT(fabs(q3 * q0 + 0.25), MAX_DIFF);
+}
+
+TEST(StaticOrientationTest, Stationary3){
```
Comment (mintar) [inline imu_filter_madgwick/test/madgwick_test.cpp]: StaticOrientationTest => MadgwickTest

**Comment by mintar** [comment]
I've finished reviewing all of it now, and let me say thanks a lot for this valuable contribution!

So let me just write down a summary (for the commit message / change log later):
1. Optionally use ENU instead of NED convention (closes #36)
2. Replace the (buggy) Euler-angle-based initialization routine (`ImuFilterRos::computeRPY`) by a (hopefully correct) transformation matrix based one (`StaticOrientation::computeOrientation`)
3. Refactor `madgwickAHRSupdate()` (pull out some functions, remove micro optimizations)
4. Add unit tests

+1 for adding the use_ned parameter and depreciation warning. Let's switch to ENU starting with L-Turtle.

+1 for replacing the initialization routine (`ImuFilterRos::computeRPY` => `StaticOrientation::computeOrientationNED`). The `computeRPY()` implementation was quite buggy. The only excuse for not fixing it earlier was that the filter would eventually recover from a bad initialization, and that fixing it was hard. :)  So thanks a lot!

+1 for pulling out the helper functions from madgwickAHRSupdate(). This makes the code much more readable.

+100 for adding unit tests for all of this! That will really make me more confident that your implementation is correct.

So, what's next? Once the code is ready from your side, just comment here and I'll do one more sanity test. Specifically, I'll test whether the new code (in NED mode) produces results that are similar to the old code on some bagfiles I've recorded. I say "similar" and not "identical" here because I expect there will be some changes due to the fixed initalization. We could put this into a "unit test" (rather regression test) as well, but for now I'll just do it manually. Then we can squash and merge the PR. Is it okay for you if I squash the PR and write the commit message (keeping your name as author of course), or do you prefer to do it yourself?

FYI, I'll be on vacation for 3 weeks starting this Saturday. So either we manage before then, or I'll merge after my vacation (no big deal).

**Comment by mistoll** [comment]
I'm glad you like it.

> So let me just write down a summary (for the commit message / change log later):
> 1.  Optionally use ENU instead of NED convention (closes #36)
> 2.  Replace the (buggy) Euler-angle-based initialization routine (ImuFilterRos::computeRPY) by a (hopefully correct) transformation matrix based one (StaticOrientation::computeOrientation)
> 3.  Refactor madgwickAHRSupdate() (pull out some functions, remove micro optimizations)
> 4.  Add unit tests

Maybe it's worth mentioning that StaticOrientation::computeOrientation is available as public library method. BTW: I don't like the name StaticOrientation do you come up with another name? Maybe StatelessOrientation?

> +100 for adding unit tests for all of this! That will really make me more confident that your implementation is correct.

Me too! However please double check the result quaternion assertions.

> So, what's next? Once the code is ready from your side, just comment here and I'll do one more sanity test. Specifically, I'll test whether the new code (in NED mode) produces results that are similar to the old code on some bagfiles I've recorded. I say "similar" and not "identical" here because I expect there will be some changes due to the fixed initalization. We could put this into a unit test as well, but for now I'll just do it manually. 

You should do that. It would be great to know that the PR doesn't break any existing systems

> Then we can squash and merge the PR. Is it okay for you if I squash the PR and write the commit message (keeping your name as author of course), or do you prefer to do it yourself?

Just do as you like. As mentioned before the commits are mixed and and should be squashed. 

> FYI, I'll be on vacation for 3 weeks starting this Saturday. So either we manage before then, or I'll merge after my vacation (no big deal).

Well, I'm done. However no need to hurry.

**Comment by mistoll** [comment]
Are you back from vacation? Ready to merge?

**Comment by mintar** [comment]
Yes, I'm back. Sorry for letting this hang for so long. I'll merge tomorrow, I promise! :)

**Comment by mintar** [comment]
Bad news: I've rebased and squashed your commits for merging, and then I tested everything extensively. Basically I wanted to make sure that this PR doesn't change the behavior when the default params are used (i.e., use_ned = true). However, it seems like that's not the case.

I've dumped the results of my testing below. I'm getting a headache from staring at orientations all day, perhaps you can make more sense of it.

My (very scientific) testing methodology consisted of placing our IMU (a Phidgets 1044) flat on a table and starting the filter in different configurations. I made sure the IMU didn't move, but of course there's some jitter in the input data that accounts for small variations in the quaternions. Anyway, I've tested with two **IMU physical orientations**: **upside down** (meaning `linear_acceleration.z = 9.77`), and **upright** (meaning `linear_acceleration.z = -9.77`). Full details below the table.

I've tested two **branches**: [ahrs_enu_rebased](https://github.com/mintar/imu_tools/tree/ahrs_enu_rebased), currently at mintar/imu_tools@7e41ec802135ead1168c935eb0cc645d389e88cb, and [indigo](https://github.com/mintar/imu_tools/tree/indigo), currently at mintar/imu_tools@3f829d90cae85c802e419049fbdc2f96e7c74b73 . The first is what I propose to merge (exactly the squashed commits in this PR, but I've renamed StaticOrientation to StatelessOrientation, as you suggested and fixed some typos in the comments). The second is basically the current indigo branch, but both branches contain a "DEBUG OUTPUT" commit (mintar/imu_tools@3f829d90cae85c802e419049fbdc2f96e7c74b73). That commit prints out the quaternions in the full (google docs) table linked below.

The **stateless orientation** column (and the quaternion after that in the full table) represents the first call to `setOrientation`, so either from the [old code](https://github.com/mintar/imu_tools/blob/7e41ec802135ead1168c935eb0cc645d389e88cb/imu_filter_madgwick/src/imu_filter_ros.cpp#L159-L167) when use_mag = false, or from [your new code](https://github.com/mintar/imu_tools/blob/7e41ec802135ead1168c935eb0cc645d389e88cb/imu_filter_madgwick/src/imu_filter_ros.cpp#L224-L230) when use_mag = true.

I've marked the bugs in the table below in bold. Some comments:
- **Bug:** When the IMU is upside down, use_ned = true, use_mag = false (row 2), the filter is initialized with z pointing up, but then the filter slowly slowly converges to z pointing down. This is because the initialization still uses the [old code](https://github.com/mintar/imu_tools/blob/7e41ec802135ead1168c935eb0cc645d389e88cb/imu_filter_madgwick/src/imu_filter_ros.cpp#L159-L167). Solution: Add a function to stateless_orientation.h that doesn't require a magnetic field and use that.
- When use_ned = false, use_mag = false (rows 4, 6, 11, 13), both indigo and ahrs_enu_rebased branches agree.
- When use_ned = false, use_mag = true (rows 5, 7, 12, 14), both indigo and ahrs_enu_rebased branches agree except that yaw(ahrs_enu_rebased) = yaw(indigo) + pi/2. I guess this pi/2 is because the old code (indigo) would set yaw = 0 when the x axis points north, and the new code (with use_ned = false, meaning ENU) would set yaw = 0 when the x axis points east. I have to check whether those two things are true. By now I'm totally confused what ENU and NED actually means, but I guess the old code was buggy and ended up unintentionally implementing something like NWU.
- **Bug:** In row 9, both the filter and the old initialization code say "z down" when use_mag = false, but "z up" when use_mag = true (row 10). No idea why this only happens when the IMU is upright, but I'm pretty sure row 9 + row 10 should only differ in yaw.

So, what do we do next? I guess a good step forward would be putting the stuff from the table into unit tests, but I'm still unsure what the correct behavior would be in most of the cases. I really have to read up on NED and ENU again.

---

# Test results

The full table, including the quaternions of the orientations, is [here](https://docs.google.com/spreadsheets/d/1jXTEf4jrrOo1lWaRwA7NJfJ_HK7aWjQbclLzCCuHYE4/edit#gid=1256043030).

| row | branch | IMU physical orientation | use_ned | use_mag | stateless orientation | filter converges to |
| --- | --- | --- | --- | --- | --- | --- |
| 2 | ahrs_enu_rebased | upside down | true | false | **up** | down |
| 3 | ahrs_enu_rebased | upside down | true | true | down | down |
| 4 | ahrs_enu_rebased | upside down | false | false | up | up |
| 5 | ahrs_enu_rebased | upside down | false | true | up, yaw = indigo + pi/2 | up, yaw = indigo + pi/2 |
| 6 | indigo | upside down |  | false | up | up |
| 7 | indigo | upside down |  | true | up | up |
| 8 |  |  |  |  |  |  |
| 9 | ahrs_enu_rebased | upright | true | false | **down** | **down** |
| 10 | ahrs_enu_rebased | upright | true | true | up | up |
| 11 | ahrs_enu_rebased | upright | false | false | down | down |
| 12 | ahrs_enu_rebased | upright | false | true | down, yaw = indigo + pi/2 | down, yaw = indigo + pi/2 |
| 13 | indigo | upright |  | false | down | down |
| 14 | indigo | upright |  | true | down | down |

# IMU physical orientation

## upside down

```
$ rostopic echo /imu/mag
header: 
  seq: 12193
  stamp: 
    secs: 1472739636
    nsecs: 161754219
  frame_id: imu
vector: 
  x: 8.484e-06
  y: 8.181e-06
  z: -4.3329e-05

$ rostopic echo /imu/data_raw
header: 
  seq: 30
  stamp: 
    secs: 1472739774
    nsecs: 861754220
  frame_id: imu
orientation: 
  x: 0.0
  y: 0.0
  z: 0.0
  w: 0.0
orientation_covariance: [-1.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0]
angular_velocity: 
  x: -0.00105452793405
  y: 0.00246073971239
  z: -0.000936892742471
angular_velocity_covariance: [1.2184696791468346e-07, 0.0, 0.0, 0.0, 1.2184696791468346e-07, 0.0, 0.0, 0.0, 1.2184696791468346e-07]
linear_acceleration: 
  x: 0.201301208611
  y: 0.555344123755
  z: 9.7791970183
linear_acceleration_covariance: [8.66124974095918e-06, 0.0, 0.0, 0.0, 8.66124974095918e-06, 0.0, 0.0, 0.0, 8.66124974095918e-06]
```

## upright

```
$ rostopic echo /imu/mag
header: 
  seq: 2439
  stamp: 
    secs: 1472741808
    nsecs: 433886858
  frame_id: imu
vector: 
  x: 6.363e-06
  y: 1.0908e-05
  z: 4.2723e-05

$ rostopic echo /imu/data_raw
header: 
  seq: 0
  stamp: 
    secs: 1472741892
    nsecs: 657886858
  frame_id: imu
orientation: 
  x: 0.0
  y: 0.0
  z: 0.0
  w: 0.0
orientation_covariance: [-1.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0]
angular_velocity: 
  x: 0.00246073971239
  y: 0.0
  z: 0.00187378548494
angular_velocity_covariance: [1.2184696791468346e-07, 0.0, 0.0, 0.0, 1.2184696791468346e-07, 0.0, 0.0, 0.0, 1.2184696791468346e-07]
linear_acceleration: 
  x: 0.129492005539
  y: 0.291847512484
  z: -9.8330539206
linear_acceleration_covariance: [8.66124974095918e-06, 0.0, 0.0, 0.0, 8.66124974095918e-06, 0.0, 0.0, 0.0, 8.66124974095918e-06]
```

**Comment by mistoll** [comment]
Well, I tried to prevent this kind of differences by implementing the unit tests. But those quaternions are real brain teasers. Here's my suggestion
1. I'm going to implement the tests you've done as unit test
2. Run those tests against the indigo branch (test the tests)
3. Run the tests against the ahrs_enu branch and fix the bugs

Going to look into this tomorrow.

**Comment by mintar** [comment]
That sounds like a good plan. Even if the current implementation is buggy, we should keep the old behavior by default.

**Comment by mistoll** [comment]
First of all: I assumed that "upright" means Z is pointing downwards (as Z measures -9,81). Hence "upside down" means Z is pointing upwards.

I'm done with (1). The values are equal for stateless and ahrs filters.

| IMU | IMU XYZ | Frame | Expected rotation | Exp. quaternion (w,x,y,z) |
| --- | --- | --- | --- | --- |
| upright | west north down | ENU | Y, 180deg | 0.0, -0.25, -0.97, -0.02 |
| upright | west north down | NED | Z, 90deg | 0.86, -0.01, 0.01, -0.50 |
| upside-down | NE NW up | ENU | Z, 45deg | 0.93, 0.03, 0.0, 0.35 |
| upside-down | NE NW up | NED | Z, 45deg + X 180 deg | 0.021, -0.91, -0.41, 0.02 |

Do you agree on the expected results for those cases?

These test succeed with ahrs_enu (3). They fail with indigo (2), but I still have to analyse that.

**Comment by mintar** [comment]
I'll reply to your comment as soon as possible (I have a meeting starting now that'll take forever), but first here's something I wrote this morning:

Okay, I've refreshed my memory of what ENU and NED actually means (it's been 1.5 years since I last dealt with this). I'm just repeating the obvious here to get it straight in my own mind. The clearest explanation I could find is in ros-infrastructure/rep#95:

> - For NED type IMUs, the orientation of the world frame is x-north, y-east, z-down, relative to magnetic north.
> - For ENU type IMUs, the orientation of the world frame is x-east, y-north, z-up, relative to magnetic north.
> - [...]
> - The device has an associated neutral orientation, defined as the orientation of the device where the body and the world frame align.

I don't like the wording "NED/ENU type IMUs". If I'm correct, the IMU device itself only produces magnetometer / accelerometer / gyro readings, and it's up to the fusion stage (imu_filter_madgwick in this case) to decide whether to output NED or ENU. With that in mind, I think the tests that you added are correct:

### ENU (test Stationary1)

```
float Ax = 0.0, Ay = 0.0, Az = 9.81;
float Mx = 0.0, My = 0.0005, Mz = -0.0005;
[...]
filter.getOrientation(q0,q1,q2,q3);
ASSERT_LT(fabs(q1), MAX_DIFF);
ASSERT_LT(fabs(q2), MAX_DIFF);
ASSERT_LT(fabs(q3), MAX_DIFF);
```

For an ENU type IMU, the neutral orientation is when Ax = 0, Ay = 0 and Az > 0 (the "Up" in ENU), Mx = 0, My > 0 (the "East" in ENU) and Mz arbitrary (negative in the northern hemisphere, positive in the southern hemisphere).

I've carefully oriented the IMU to get these values:

```
mag:
  x:  0.000000000000000000
  y:  0.000010605000000000
  z: -0.00004363200000000

linear_acceleration: 
  x: 0.193060808258
  y: 0.580752024841
  z: 9.81735791993
```

... which results in the following orientations:

**branch "indigo"**

```
orientation:
  x: -0.0151375354836
  y: 0.0273694787413
  z: 0.682634575192
  w: -0.727775386177
```

In RPY, this is roughly (0, 0, -pi/2).

**branch "ahrs_enu_rebased", use_ned = false**

neutral orientation:

```
orientation: 
  x: 0.0301468137945
  y: -0.00857729893563
  z: 0.0333262765768
  w: 0.99726352394
```

In RPY, this is roughly (0, 0, 0).

**branch "ahrs_enu_rebased", use_ned = true**

```
orientation: 
  x: -0.727326172442
  y: -0.683135511796
  z: 0.0271042773862
  w: 0.0146026759666
```

In RPY, this is roughly (-pi, 0, pi/2).

### NED (test StationaryNED1)

```
float Ax = 0.0, Ay = 0.0, Az = -9.81;
float Mx =  0.0005, My = 0.0, Mz = 0.0005;
[...]
filter.getOrientation(q0,q1,q2,q3);
ASSERT_LT(fabs(q1), MAX_DIFF);
ASSERT_LT(fabs(q2), MAX_DIFF);
ASSERT_LT(fabs(q3), MAX_DIFF);
```

For a NED type IMU, the neutral orientation is when Ax = 0, Ay = 0 and Az < 0 (the "Down" in NED), Mx > 0, My = 0 (the "Down" in NED) and Mz arbitrary (positive in the northern hemisphere, negative in the southern hemisphere).

### neutral orientations (quaternion x=0, y=0, z=0, w= +/- 1)

**branch indigo**

```
linear_acceleration: 
  x: 0.196101908388
  y: 0.569568624363
  z: 9.81804461996

mag: 
  x:  0.000013938000000000
  y: -0.00000242400000000
  z: -0.00004393500000000
```

(strangely, if I make mag.y = 0, the filtered output is not exactly the neutral orientation any more).

So in terms of the tests, that would be:

```
float Ax = 0.0, Ay = 0.0, Az = 9.81;
float Mx =  0.0005, My = 0.0, Mz = -0.0005;
```

In other words, the current code (indigo) publishes NWU. :(

**Comment by mistoll** [comment]
NWU, that matches what I'm seeing. But that's not too bad. We will have to switch between NWU, NED and ENU.

I will add test cases for NWU and see what the indigo branch does with it.

Is it ok for you, if I force push a new set of commits to this pr? I'd like to have the unit tests for the current code in a separate commit.

**Comment by mintar** [comment]
> Is it ok for you, if I force push a new set of commits to this pr? I'd like to have the unit tests for the current code in a separate commit.

Sure! It would be great if you could base your stuff on the two commits in the mintar/imu_tools ahrs_enu_rebased branch. Feel free to completely take it apart, rearrange commits and whatnot. All I'm saying is that the ahrs_enu_rebased branch is the most recent version of this code (includes the rename static => stateless).

**Comment by mintar** [comment]
Re NWU/NED/ENU : there is a static transformation between all of those. So in order to avoid code duplication, we could pick one of them to work internally (say, ENU) and simply optionally apply a fixed rotation at the end to convert to NED / NWU.

**Comment by mistoll** [comment]
I've implemented a NWU/NED/ENU switch. The places where they differ are very straightforward.

So let's see what we've got at https://github.com/ccny-ros-pkg/imu_tools/pull/60/commits/c331633c83857963051ae0db8e4ca60ceb400836
- Unit tests for NWU, NED, ENU
- All tests succeed on ahrs_enu branch
- All NWU tests succeed on current indigo branch
- Parameters to switch between NWU, NED, ENU

Not covered by tests:
- Dynamics (Gyro)
- AHRS Filter without magnetic field
- ROS interface

Refering to the test results:
- row 2: should be fixed, Not covered by test
- row 5 and 12: NWU issue, should be fixed
- row 9 stateless: should be fixed, same as row 2
- row 9 ahrs: couldn't figure that out. Not covered by test

I thing the filter with mag should be good now. For the rest unit tests would be great. However the quaternions are not unique in those cases. Thus the tests are harder to validate. I'll see what I can do.

Could  you please verify the results for the case where mag is enabled?

**Comment by mintar** [comment]
From what I can tell, those results look good to me. I cannot be sure though; it's always hard to imagine the orientations in one's head.

I've done one more thing. I've recorded a rosbag of our IMU turning through as many orientations as I could manage; at least with the magnetic vector, I've covered the full sphere.

The topic "/imu/data_nwu" contains the filtered output of the current indigo branch.

You can test it like this:
- Check out your current `ahrs_enu` branch, compile.
- Then:
  
  wget http://kos.informatik.uos.de/rosbags/imu_tools/phidgets_imu_nwu_full_sphere.bag
  roscore
  rosparam set use_sim_time true
  rosrun imu_filter_madgwick imu_filter_node _use_mag:=true _use_ned:=false _use_nwu:=true
  rosbag play --clock phidgets_imu_nwu_full_sphere.bag
- Now, open up RViz, select `odom` as the fixed frame and add two `rviz_imu_plugin/Imu` visualizations, one on topic "/imu/data" and one on "/imu/data_nwu". Only enable the axes.

Now you should be able to visually compare your current code with the old code.

My results show that NWU mode seems to be pretty flawless. The IMU starts off at a slight offset (which is to be expected due to different initialization) but eventually converges close to the indigo branch.

ENU mode seems to have problems. It should be a fixed 90° rotation to NWU, but it isn't; sometimes the wrong two axes line up.

NED mode seems to be completely borked. It should also be a fixed rotation to NWU (right?), but it seems to be rotating in the opposite direction as NWU. What really confuses me about this one is when you compare the initial 11 seconds, NWU doesn't move at all but NED does (and I'm pretty sure I rotated the IMU). I guess I must have aligned some axis funnily.

**Comment by mistoll** [comment]
Now we've got unit tests for the filters without magnetic fields, too. However I'm not sure if the logic behind the quat_eq_ex_z method is correct. It should compare the quaternions but ignore any difference in yaw. Well, there's a unit test for that method.

The test values with or without magnetic field are the same. So if row 9 and 10 still differ, there's something wrong in the ROS interface. AFAIK it's possible to unit test that, too. However I've never done it before. And the logic there is quite simple, compared to the filters. Let's omit those.

The bag file is great. However it's difficult to separate concerns. Especially static and dynamic. If you have the chance to record another bag file, please add some long static phases. 
Just had a short look, though. Going to do some more analysis next week.

**Comment by mistoll** [comment]
I've added a "stateless" switch. With stateless=true the filter does only use the stateless orientation. That's primarily for debug purpose.

After staring at rviz for some time, I simply inverted all the orientations, using another node. Now they match pretty good. 

[Invertion filter](https://www.dropbox.com/s/nsu8wdavf6gibw0/imu_orientation_filter.py?dl=0)

How to reproduce:
- Start imu_filter_madwick
- Run imu_orientation_filter.py imu:=/imu/data imu_filtered:=/imu/data2
- Run imu_orientation_filter.py imu:=/imu/data_nwu imu_filtered:=/imu/data_nwu2 __name:=f
- Start rosbag
- Display /imu/data2 and /imu/data_nwu2 in rviz

Interpretation:

The orientation in the imu message is the orientation of the imu relative to the earth. When inverting that it is the earth relative to the imu. With ENU, NWU and NED we switch the frame of the earth. So the earth frames are fixed to each other. 

Hence the filter is working correctly. It takes some time to converge sometimes.

**Comment by mintar** [comment]
Okay, that last point convinced me. I've also spent some more time staring at rviz. Then I came up with another good way to verify that the filter is working correctly:
- Make three earth frames in tf. NWU -> ENU is yaw -pi/2, NWU ->NED is roll +pi. (static_transform_publisher's arguments are x y z _yaw pitch roll_, not roll pitch yaw).
  
  ```
  rosrun tf static_transform_publisher 0 0 0 -1.57 0 0 earth_nwu earth_enu 10
  rosrun tf static_transform_publisher 0 0 0 0 0 3.14 earth_nwu earth_ned 10
  ```
- Rename imu_frame and fixed_frame based on whether we set use_ned/use_enu (only for this test, in mintar/imu_tools@9fe46555e29a3d324c74304abfecff92aee03d59).
- Start three instances of imu_filter_madgwick, one for NWU/NED/ENU each:
  
  ```
  rosrun imu_filter_madgwick imu_filter_node _publish_tf:=true _use_mag:=true _use_ned:=false _use_nwu:=true  /imu/data:=/imu/data_nwu /imu/magnetic_field:=/imu/magnetic_field_nwu __name:=imu_nwu
  rosrun imu_filter_madgwick imu_filter_node _publish_tf:=true _use_mag:=true _use_ned:=false _use_nwu:=false /imu/data:=/imu/data_enu /imu/magnetic_field:=/imu/magnetic_field_enu __name:=imu_enu
  rosrun imu_filter_madgwick imu_filter_node _publish_tf:=true _use_mag:=true _use_ned:=true  _use_nwu:=false /imu/data:=/imu/data_ned /imu/magnetic_field:=/imu/magnetic_field_ned __name:=imu_ned
  ```
- The tf tree now looks like this:
  frames
- Open up RViz, _don't_ use the Imu plugin, instead select "earth_nwu" (or any other earth frame) as fixed frame and visualize the tf tree.

**TADAAA**, all three frames (imu_nwu, imu_ned, imu_enu) align! So now I'm convinced the filter works correctly. :)

**Comment by mintar** [comment]
I think I'm going to keep your "stateless" param. It's quite useful for debugging, and I've manually made that change in the code during debugging several times now because that param was missing.

I think this is good to merge. If you don't mind, I'm going to replace the use_ned/use_enu boolean params by a single earth_frame param (mintar/imu_tools@88f86fb). It's a string param, and I don't usually like them a lot, but I think in this case it's the best option.

So if you don't protest loudly by tomorrow I'll merge and do a new release. Thanks again for all the hard work you've put in, I think the result is fantastic! :)

**Comment by mistoll** [comment]
The stateless param could also help debugging IMU issues, because in statless mode the gyro is ignored. So keep it!

I don't like having two booleans to switch between three modes. IMO a string param is a much better user experience. But please confirm, that the name "earth_frame" is the correct term.

Go ahead, merge it!

**Comment by mintar** [comment]
Good point. The upcoming REP 145 (ros-infrastructure/rep#95) calls it "world frame", so I'm going to rename earth frame to world frame and then merge.

**Comment by mintar** [comment]
All merged and released. Thanks again! :)
